src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs

```diff
@@ -0,0 +1,252 @@
+
+            // Start at slot 1 (slot 0 just indicates reachability)
+            for (int slot = 1; slot < state.Capacity; slot++)
+            {
+                if (state[slot])
+                {
+                    if (isWrite)
+                    {
+                        SetSlotAssigned(slot);
+                    }
+                    else
+                    {
+                        var symbol = variableBySlot[slot].Symbol;
+                        CheckAssigned(symbol, syntax, slot);
+                    }
+                }
+            }
+
+            usages.LocalFuncVisited = true;
+        }
+
+        private int RootSlot(int slot)
+        {
+            while (true)
+            {
+                var varInfo = variableBySlot[slot];
+                if (varInfo.ContainingSlot == 0)
+                {
+                    return slot;
+                }
+                else
+                {
+                    slot = varInfo.ContainingSlot;
+                }
+            }
+        }
+
+        public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatement localFunc)
+        {
+            var oldMethodOrLambda = this.currentMethodOrLambda;
+            this.currentMethodOrLambda = localFunc.Symbol;
+
+            var oldPending = SavePending(); // we do not support branches into a lambda
+
+            // Local functions don't affect outer state and are analyzed
+            // with everything unassigned and reachable
+            var savedState = this.State;
+            this.State = this.ReachableState();
+
+            var usages = GetOrCreateLocalFuncUsages(localFunc.Symbol);
+            var oldReads = usages.ReadVars;
+            usages.ReadVars = BitVector.Empty;
+
+            if (!localFunc.WasCompilerGenerated) EnterParameters(localFunc.Symbol.Parameters);
+
+            var oldPending2 = SavePending();
+            VisitAlways(localFunc.Body);
+            RestorePending(oldPending2); // process any forward branches within the lambda body
+            ImmutableArray<PendingBranch> pendingReturns = RemoveReturns();
+            RestorePending(oldPending);
+            LeaveParameters(localFunc.Symbol.Parameters, localFunc.Syntax, null);
+
+            LocalState stateAtReturn = this.State;
+            foreach (PendingBranch pending in pendingReturns)
+            {
+                this.State = pending.State;
+                if (pending.Branch.Kind == BoundKind.ReturnStatement)
+                {
+                    // ensure out parameters are definitely assigned at each return
+                    LeaveParameters(localFunc.Symbol.Parameters, pending.Branch.Syntax, null);
+                }
+                else
+                {
+                    // other ways of branching out of a lambda are errors, previously reported in control-flow analysis
+                }
+
+                IntersectWith(ref stateAtReturn, ref this.State);
+            }
```

Comment:
Please move this inside the "if" branch. It looks lonely out here ;)

Also, it should only be done for return statements.